### ISSUE TITLE
Static constructors: false positive for `no-member`

### ISSUE THREAD

**Issue by torotil** [PyCQA/pylint#11119]
### Bug description

```python
Type inference doesn’t seem to work correctly when inferring the type returned by a static constructor.
```

### Configuration

```ini
# Using the default configuration
```

### Command used

```shell
pylint no_member.py
```

### Pylint output

```python
************* Module no_member
no_member.py:6:0: R0903: Too few public methods (1/2) (too-few-public-methods)
no_member.py:20:7: E1101: Instance of 'A' has no 'new_method' member (no-member)

------------------------------------------------------------------
Your code has been rated at 4.55/10 (previous run: 4.55/10, +0.00)
```

### Expected behavior

The `no-member` error should not be triggered because the type of `B.create()` is actually `B` not `A`.

[no_member.py](https://github.com/user-attachments/files/29382728/no_member.py)

### Pylint version

```shell
pylint 4.0.6
astroid 4.0.4
Python 3.13.14 (main, Jun 10 2026, 18:10:12) [GCC 15.2.0]
```

### Additional dependencies

```python

```

**Comment by clavedeluna** [comment]
reproduced

**Comment by utkarshalpha** [comment]
Some root-cause detail, since this reproduces purely at the astroid inference layer and does not involve any pylint checker logic.

With pylint 4.0.6 / astroid 4.0.4:

```python
import astroid
mod = astroid.parse(open("no_member.py").read())
A = mod.locals['A'][0]
next(A.locals['create'][0].body[0].value.infer())   # cls()  -> Instance of .A
```

`cls()` inside `A.create` infers to `Instance of A` rather than to the class the method was actually called on, so the receiver is not propagated. That result then flows outward unchanged:

| expression | inferred |
|---|---|
| `cls()` in `A.create` | `Instance of A` |
| `super().create()` in `B.create` | `Instance of A` |
| `B.create()` | `Instance of A` |

Since `B.create()` ends up as an `A` instance, `no-member` is correct given its input; the wrong answer is produced upstream. The `-> t.Self` annotations are not consulted, so the same false positive appears without them.

That suggests the fix belongs in astroid's bound-classmethod inference (propagating the actual receiver class through `cls`, and/or honoring `typing.Self` as a return annotation) rather than in the `no-member` checker. Filing here since this issue is the visible symptom; happy to move it to the astroid tracker if you prefer.

**Comment by torotil** [comment]
This seems to be already covered by https://github.com/pylint-dev/astroid/issues/2852 then.
